pkg/operator/ceph/cluster/controller.go

```diff
@@ -256,81 +256,57 @@ func (c *ClusterController) onAdd(obj interface{}) {
 	logger.Infof("starting cluster in namespace %s", cluster.Namespace)
 
 	if c.devicesInUse && cluster.Spec.Storage.AnyUseAllDevices() {
-		message := "using all devices in more than one namespace not supported"
-		logger.Error(message)
-		if err := c.updateClusterStatus(clusterObj.Namespace, clusterObj.Name, cephv1.ClusterStateError, message); err != nil {
-			logger.Errorf("failed to update cluster status in namespace %s: %+v", cluster.Namespace, err)
-		}
+		c.updateClusterStatus(clusterObj.Namespace, clusterObj.Name, cephv1.ClusterStateError, "using all devices in more than one namespace not supported")
 		return
 	}
 
 	if cluster.Spec.Storage.AnyUseAllDevices() {
 		c.devicesInUse = true
 	}
 
-	if cluster.Spec.Mon.Count <= 0 {
-		logger.Warningf("mon count is 0 or less, should be at least 1, will use default value of %d", mon.DefaultMonCount)
-		cluster.Spec.Mon.Count = mon.DefaultMonCount
-		cluster.Spec.Mon.AllowMultiplePerNode = true
-	}
-	if cluster.Spec.Mon.Count > mon.MaxMonCount {
-		logger.Warningf("mon count is bigger than %d (given: %d), not supported, changing to %d", mon.MaxMonCount, cluster.Spec.Mon.Count, mon.MaxMonCount)
-		cluster.Spec.Mon.Count = mon.MaxMonCount
-	}
+	c.initializeCluster(cluster, clusterObj)
+}
+
+func (c *ClusterController) initializeCluster(cluster *cluster, clusterObj *cephv1.CephCluster) {
+	cluster.Spec = &clusterObj.Spec
 	if cluster.Spec.Mon.Count%2 == 0 {
 		logger.Warningf("mon count is even (given: %d), should be uneven, continuing", cluster.Spec.Mon.Count)
 	}
 
 	// Start the Rook cluster components. Retry several times in case of failure.
-	validOrchestration := true
-	err = wait.Poll(clusterCreateInterval, clusterCreateTimeout, func() (bool, error) {
-		cephVersion, err := cluster.detectCephVersion(c.rookImage, cluster.Spec.CephVersion.Image, detectCephVersionTimeout)
-		if err != nil {
-			logger.Errorf("unknown ceph major version. %+v", err)
-			return false, nil
-		}
+	failedMessage := ""
+	state := cephv1.ClusterStateError
 
-		if !cluster.Spec.CephVersion.AllowUnsupported {
-			if !cephVersion.Supported() {
-				err = fmt.Errorf("unsupported ceph version detected: %s. allowUnsupported must be set to true to run with this version", cephVersion)
-				logger.Errorf("%+v", err)
-				validOrchestration = false
-				// it may seem strange to log error and exit true but we don't want to retry if the version is not supported
+	err := wait.Poll(clusterCreateInterval, clusterCreateTimeout, func() (bool, error) {
```

Comment:
I found myself confused for quite a while trying to figure out how a return-less function is returning `bool, err`. It's very easy to miss that this comes from this function definition as a parameter. Could we change this so that the `func() (bool, error)` is defined before the `wait.Poll` call? It also seems just as good/better to define a new function for this `func (c *clusterController) funcName() (bool, err)` and send the `failedMessage` with the error when returning.

Reply:
While readability of inline methods is always going to have its challenges, I think defining it like this is also helpful because it's only used in this one place and you can easily see that context that it's for continued polling. In fact, everywhere rook calls `wait.Poll()`, it defines the method inline.

Reply:
I've encountered this a few times in the past week, and I get tripped up every time. Can we meet in the middle and move the `func() (bool, error) {` to the next line so that it stands out more? The examples I see of go code online with functions as parameters do this, and those functions are seldom more than 10 lines long.